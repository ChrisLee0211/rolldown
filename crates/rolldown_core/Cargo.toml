[package]
name = "rolldown_core"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rolldown_common = { path = "../rolldown_common" }
rolldown_swc_visitors = { path = "../rolldown_swc_visitors" }
rolldown_compiler = { path = "../rolldown_compiler" }
rolldown_resolver = { path = "../rolldown_resolver" }
rolldown_error = { path = "../rolldown_error" }
rolldown_runtime_helpers = { path = "../rolldown_runtime_helpers" }
rolldown_ast_template = { path = "../rolldown_ast_template" }
rolldown_plugin = { path = "../rolldown_plugin" }
rolldown_tracing = { path = "../rolldown_tracing" }
tokio = { workspace = true, features = ["full"] }
swc_core = { workspace = true, features = [
    "ecma_ast",
    "common",
    "common_concurrent",
    "ecma_utils",
] }
swc_node_comments = { workspace = true }
derivative = { workspace = true }
once_cell = { workspace = true }
hashlink = { workspace = true }
sugar_path = { workspace = true }
tracing = { workspace = true }
itertools = { workspace = true }
rustc-hash = { workspace = true }
rayon = { workspace = true }
futures = {  workspace = true }
phf = { workspace = true , features = ["macros"] }
regex = "1.5.4"

[dev_dependencies]
rolldown_test_utils = { path = "../rolldown_test_utils" }
insta = { workspace = true }
testing_macros = { workspace = true }
criterion = "0.4.0"

[[bench]]
name = "threejs"
harness = false