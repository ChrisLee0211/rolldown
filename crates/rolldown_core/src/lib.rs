#![feature(map_many_mut)]
#![feature(box_syntax)]
#![feature(hash_raw_entry)]
#![feature(iterator_try_collect)]
use std::sync::Arc;

mod bundler;
pub use bundler::*;
mod chunk;
pub use chunk::*;
mod normal_module;
pub use normal_module::*;
mod external_module;
pub use external_module::*;
mod options;
pub use options::*;
mod graph;
pub use graph::*;
mod module_loader;
use rolldown_common::{ChunkId, ExportedSpecifier, ModuleId};
use rolldown_resolver::Resolver;
use rustc_hash::FxHashMap;
use swc_core::common::{FilePathMapping, Globals, SourceMap};
mod bundle;
mod norm_or_ext;
pub use bundle::*;
use swc_core::ecma::atoms::JsWord;
mod code_splitter;
pub use code_splitter::*;
mod chunk_graph;
pub(crate) use chunk_graph::*;
mod plugin_driver;
pub(crate) use plugin_driver::*;
mod utils;
pub use utils::*;
mod rolldown_output;
mod treeshake;
pub use rolldown_output::*;

pub(crate) type ResolvedModuleIds = FxHashMap<JsWord, ModuleId>;
pub(crate) type MergedExports = FxHashMap<JsWord, ExportedSpecifier>;
pub(crate) type SharedResolver = Arc<Resolver>;
pub(crate) use norm_or_ext::*;
use once_cell::sync::Lazy;

pub(crate) static SOURCE_MAP: Lazy<Arc<SourceMap>> =
  Lazy::new(|| Arc::new(SourceMap::new(FilePathMapping::empty())));

pub(crate) static COMPILER: Lazy<Arc<rolldown_compiler::Compiler>> = Lazy::new(|| {
  let cm = SOURCE_MAP.clone();
  let compiler = rolldown_compiler::Compiler::with_cm(cm);
  Arc::new(compiler)
});

pub(crate) type ModuleById = FxHashMap<ModuleId, NormOrExt>;
pub(crate) type ModuleRefMutById<'a> = FxHashMap<&'a ModuleId, &'a mut NormOrExt>;
pub(crate) type SplitPointIdToChunkId = FxHashMap<ModuleId, ChunkId>;
pub(crate) static SWC_GLOBALS: Lazy<Arc<Globals>> = Lazy::new(|| Arc::new(Globals::new()));

// public exports

pub type BundleResult<T> = rolldown_error::Result<T>;
pub type BundleError = rolldown_error::Error;
