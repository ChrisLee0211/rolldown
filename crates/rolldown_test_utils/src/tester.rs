use std::{
  collections::HashSet,
  path::{Path, PathBuf},
  sync::Arc,
};

use futures::FutureExt;
use rolldown_core::BuildError;

use crate::test_config::TestConfig;

pub struct Tester {
  pub config: TestConfig,
  pub warnings: Vec<BuildError>,
}

impl Tester {
  pub fn from_config_path(filepath: &Path) -> Self {
    let test_config = TestConfig::from_config_path(filepath);
    Self {
      config: test_config,
      warnings: Default::default(),
    }
  }

  pub fn input_options(&self, cwd: PathBuf) -> rolldown_core::InputOptions {
    rolldown_core::InputOptions {
      input: self.config.input.input.clone(),
      cwd,
      treeshake: self.config.input.treeshake,
      is_external: {
        let external = self
          .config
          .input
          .external
          .clone()
          .into_iter()
          .collect::<HashSet<_>>();
        Arc::new(move |specifier, _importer, _| {
          let external = external.clone();
          futures::future::ready(Ok(external.contains(specifier))).boxed()
        })
      },
      ..Default::default()
    }
  }
}
