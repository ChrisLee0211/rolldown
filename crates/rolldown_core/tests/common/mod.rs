use std::{path::Path, str::FromStr};

use rolldown_core::{Asset, BuildResult, Bundler, ExportMode, InternalModuleFormat, OutputOptions};
use rolldown_test_utils::tester::Tester;

pub struct CompiledFixture {
  pub tester: Tester,
  pub bundler: Bundler,
  pub output: BuildResult<Vec<Asset>>,
  pub name: String,
}

impl CompiledFixture {
  pub fn output_friendly_to_snapshot(&self) -> String {
    let mut assets = self.output.as_ref().unwrap().iter().collect::<Vec<_>>();
    assets.sort_by_key(|c| &c.filename);
    assets
      .iter()
      .flat_map(|asset| {
        [
          format!("---------- {} ----------", asset.filename),
          asset.content.trim().to_string(),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n")
  }
}

pub async fn compile_fixture(test_config_path: &Path) -> CompiledFixture {
  let fixture_path = test_config_path.parent().unwrap();

  let tester = Tester::from_config_path(test_config_path);

  let mut bundler = Bundler::new(tester.input_options(fixture_path.to_path_buf()));

  if fixture_path.join("dist").is_dir() {
    std::fs::remove_dir_all(fixture_path.join("dist")).unwrap();
  }

  let output = bundler
    .generate(OutputOptions {
      // dir: Some(fixture_path.join("dist").to_string_lossy().to_string()),
      format: InternalModuleFormat::from_str(&tester.config.output.format).unwrap(),
      export_mode: ExportMode::from_str(&tester.config.output.export_mode).unwrap(),
      ..Default::default()
    })
    .await;
  let fixture_name = fixture_path
    .file_name()
    .unwrap()
    .to_string_lossy()
    .to_string();

  CompiledFixture {
    tester,
    bundler,
    output,
    name: fixture_name,
  }
}

pub fn run_test(test_config_path: &Path) {
  // compile the fixture folder
  let compiled_fx = tokio::runtime::Runtime::new()
    .unwrap()
    .block_on(crate::common::compile_fixture(test_config_path));

  // If the test config has an expected error, assert that the error matches
  if let Some(expected_error) = compiled_fx.tester.config.expected_error {
    let error = compiled_fx
      .output
      .expect_err("Expected error but got success");
    assert_eq!(error.kind.code(), expected_error.code);
    assert_eq!(error.kind.to_string(), expected_error.message);
    return;
  }

  // Otherwise, assert that the output matches the snapshot

  // Configure insta to use the test config path as the snapshot path
  let fixture_folder = test_config_path.parent().unwrap();
  let mut settings = insta::Settings::clone_current();
  settings.set_snapshot_path(fixture_folder);
  settings.set_prepend_module_to_snapshot(false);
  settings.set_input_file(fixture_folder);
  settings.bind(|| {
    insta::assert_snapshot!("output", compiled_fx.output_friendly_to_snapshot());
  });
}
