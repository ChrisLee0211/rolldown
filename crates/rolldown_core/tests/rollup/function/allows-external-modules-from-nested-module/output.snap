---
source: crates/rolldown_core/tests/common/mod.rs
expression: res.content_for_snapshot()
input_file: crates/rolldown_core/tests/rollup/samples/allows-external-modules-from-nested-module
---
---------- main.js ----------
// foo.js
var path$1 = 'a/b/c';
var path2$1 = 'a/c/b';
var foo = relative(path$1, path2$1);

// main.js
import { normalize, relative } from "path";
var path = 'foo/bar/baz';
var path2 = 'foo/baz/bar';
assert.equal(relative(path, path2), normalize('../../baz/bar'));
assert.equal(foo, normalize('../../c/b'));
